### PR TITLE
Add IdentityCache support for parent-child relationship

### PR DESCRIPTION
**Pull request by patrickod**
IdentityCache provides an easy to use caching layer on top of ActiveRecord
associations, creating instance methods on models with a fetch_ prefix denoting
the use of a cache. For example in our documentation case adding IdentityCache
to Countries with `cache_has_many :presidents` would create a
`fetch_presidents` method which uses the cache instead of directly querying the
SQL table.

We allow the use of this cache by calling `.with_identity_cache` on any
`has_active` or `has_uniquely_active` relationship where the parent class
includes IdentityCache and exposes these caching methods.

### CODE REVIEW

File: lib/state_of_the_nation.rb [hunk 1]
```diff
@@ -37,17 +37,32 @@ def until(finish_key)
     end
 
     def has_active(association_plural)
-      add_child_methods(plural: association_plural, single: false)
+      @association_plural = association_plural
```
Comment (steobrien) [inline lib/state_of_the_nation.rb]: why is this an instance variable?

Comment (steobrien) [inline lib/state_of_the_nation.rb]: oh, for the internal method to see it, gotcha

File: spec/state_of_the_nation_spec.rb [hunk 2]
```diff
@@ -341,4 +341,35 @@ class Country < ActiveRecord::Base
...
   end
+
+  context "IdentityCache support" do
+    before do
+      # Mock up of the IdentityCache methods
+      module IdentityCache
+        def fetch_presidents
+          self.presidents
+        end
+      end
```
Comment (steobrien) [inline spec/state_of_the_nation_spec.rb]: Consider not mocking here?

File: spec/state_of_the_nation_spec.rb [hunk 3]
```diff
@@ -341,4 +341,35 @@ class Country < ActiveRecord::Base
...
+    end
+
+    let(:country) { Country.create }
+    let!(:president) { country.presidents.create!(entered_office_at: day(1), left_office_at: day(8)) }
+
+    context ".using_identity_cache" do
+      it "uses the fetch_method to retrieve records" do
+        expect(country).to receive(:fetch_presidents)
+          .and_call_original
+        expect(country.active_president(day(7))).to eq(president)
```
Comment (steobrien) [inline spec/state_of_the_nation_spec.rb]: this isn't a good test – if line 360 didn't include `with_identity_cache` it'd still pass

Comment (steobrien) [inline spec/state_of_the_nation_spec.rb]: I'm dumb, deleting that dumb comment

**Comment by steobrien** [comment]
Fantastic! This is a great improvement, much more explicit.

File: lib/state_of_the_nation.rb (hunk 1 above)
Comment (patrickod) [inline lib/state_of_the_nation.rb]: yep. exactly it.

**Pull request merged**